### ISSUE TITLE
Someone in javalib InetAddress clan is revealing an internal-use-only IPv4 address

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#2999]
Arman reported an instance in epollcat where a "000000000000:127.0.0.1" (really, the all binary equivalent,
the key part is the leading _12_ zeros) address is appearing and causing a equivalence test to fail.

The problem is probably in the call to get an address returned by `accept()` but it might be a case
where `equals` needs to be more inclusive (need knowledge of not just ::FF:127.0.0.1 but also of
::127.0.0.1.

Distinctions only a mother could love...

**Comment by armanbilge** [comment]
@LeeTibbert thanks for opening the issue. Truthfully it's not entirely obvious to me that this is a bug in Scala Native, Javalib, or `InetAddress`.

In https://github.com/armanbilge/epollcat/commit/8afb7bec5399853600d863a1053c626afde9a024 I added debug logging that directly prints the output of `getsockname`. The raw bytes returned are:
```scala
Array(28, 30, -9, -51, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1, 0, 0, 0, 0)
```
i.e. as far as I can tell the system itself is reporting this as the address of the socket.

**Comment by LeeTibbert** [comment]
Thank you for the additional info Arman. That helps narrow things down.

The "0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1" portion is an entirely reasonable thing for 
C `getsockname()` to have returned.  It means an IPv4 (local loopback) address reported in an IPv6 length buffer.
I think the first layer of defect is that these 16 bytes are getting stored (by "my" code) into an InetAddress(or 
subclass) and the length is not being set to 4 bytes (indicating look only at the trailing 4 bytes, i.e. IPv4 
and not a fancy ::FF: IPv6 compatible IPv4 address.

I also want to look at InetAddress* `equals` to see what they are doing.

Much as I would like to sleeze out, looks like I missed a defect.

**Comment by armanbilge** [comment]
Yes, that's what I thought as well. But I believe the second byte `30` is the `sin6_family `.
https://www.freebsd.org/cgi/man.cgi?query=inet6

The first byte `28` correctly is the `sin6_len`.

Checking the Apple headers:
```
#define	AF_INET6	30		/* IPv6 */
```
https://opensource.apple.com/source/xnu/xnu-201/bsd/sys/socket.h

It seems to me like that is being reported, and should be interpreted, as a genuine IPv6 address.

**Comment by armanbilge** [comment]
Also, check out Scala JVM behavior for the same array.
```
scala> java.net.InetAddress.getByAddress(Array(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1))
val res1: java.net.InetAddress = /0:0:0:0:0:0:7f00:1
```

**Comment by LeeTibbert** [comment]
https://docs.oracle.com/cd/E19683-01/817-0573/transition-4/index.html

the 12-zero form is IPv4–compatible address, which is valid but has
a narrowly defined utility, which does not match what we need, hence
the defect.

After we get epollcat settled down, I am going to have to write one or more
test cases to see what javalib InetAddress does. It looks like it handles
the ::FFFF: case (ipv4 mapped) but also will let a 12-zero slip through
if it ever encounters it.  Hmmmm.

**Comment by LeeTibbert** [comment]
> java.net.InetAddress.getByAddress(Array(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1))

That is expected,  for numeric addresses, Java just stuffs the supplied Array away as one
of its constructor args.  I do not remember if it clones/copies it or stores the Array itself (hence,
can change Array).

This is going to take awhile (days).

**Comment by armanbilge** [comment]
> What really has me puzzled is why it is intermittent and why macOS only.

Yes, same. Thanks for looking into it with me, I really appreciate the help on this one.

**Comment by armanbilge** [comment]
Sorry, I don't know why I didn't think to try it before: reverting epollcat back to Scala Native 0.4.7 seems to make the issue disappear. So it's something in the 0.4.8 update, that caused this issue to emerge. Which suggests you were justified in opening this issue after all 😅

**Comment by LeeTibbert** [comment]
The InetAddress changes when into 0.4.8.  0.4.9 is about to ship.  Do we want to ask Wojciech to revert the InetAddress changes?  I do not know how hard that would be.

**Comment by armanbilge** [comment]
I posted some additional notes in https://github.com/armanbilge/epollcat/issues/88#issuecomment-1316248976.

**Comment by LeeTibbert** [comment]
Update:   

This appears to be a "quirk" in macOS  which was revealed by a fix in SN 0.4.8 which
changed name resolution to follow Java behavior, particularly with respect to defaulting
to IPv4 addresses. 

Previously, on systems which had "localhost" defined as "::1" an IpV6 address would be
returned.  Now an IPv4 mapped IPv6 address is used internally. That will display as
an IPv4 address.

I hope to create and submit  trial fix for epollcat.

I checked the SN code base. SN does not use `getsockname()` in the way which triggered
the epollcat test failures.  Nothing to fix there, although I may study InetAddress more closely
for how it handles (or not, I suspect not) a "should never happen" situation (IPv4 compatible).

**Comment by LeeTibbert** [comment]
Fundamental 'quirk' appeared to be on macOS only with the first call to `gethostname()` after
an executable starts.  It returned an IPv4 compatible IPv6 address and not the expected
IPv6 mapped address.   

Scala Native does not use `gethostname()` in the way that allowed epollcat to discover
the quirk, so no changes to SN needed for this (so far!).

A PR was submitted to epollcat so that this issue is no longer relevant there.
